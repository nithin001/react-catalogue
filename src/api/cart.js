import instance from './base';
import Ajv from 'ajv';

export const get = () => {
  return new Promise((resolve, reject) => {
    instance.get('/cart')
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(get_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

export const put = (request) => {
  return new Promise((resolve, reject) => {
    instance.put('/cart', { data: request })
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(put_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

const get_schema = {
  '$id': 'http://challenge.monoqi.net/cart',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          }
        },
        'required': ['sku', 'quantity']
      }
    }
  },
  'required': ['lines']
};

const put_schema = {
  '$id': 'http://example.com/example.json',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'name': {
            '$id': '/properties/lines/items/properties/name',
            'type': 'string',
            'title': 'The Name Schema ',
            'default': '',
            'examples': [
              'IPad'
            ]
          },
          'price': {
            '$id': '/properties/lines/items/properties/price',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/price/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  229
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/price/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          },
          'lineTotal': {
            '$id': '/properties/lines/items/properties/lineTotal',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  687
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          }
        },
        'required': [
          'sku',
          'name',
          'price',
          'quantity',
          'lineTotal'
        ]
      }
    },
    'total': {
      '$id': '/properties/total',
      'type': 'object',
      'properties': {
        'amount': {
          '$id': '/properties/total/properties/amount',
          'type': 'integer',
          'title': 'The Amount Schema ',
          'default': 0,
          'examples': [
            1126
          ]
        },
        'currency': {
          '$id': '/properties/total/properties/currency',
          'type': 'string',
          'title': 'The Currency Schema ',
          'default': '',
          'examples': [
            'EUR'
          ]
        }
      },
      'required': [
        'amount',
        'currency'
      ]
    }
  },
  'required': [
    'lines',
    'total'
  ]
};