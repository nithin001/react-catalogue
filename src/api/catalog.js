import instance from './base';
import Ajv from 'ajv';

export const get = () => new Promise((resolve, reject) => {
  instance.get('/catalog')
    .then((response) => {
      if (response.data === undefined) {
        reject('backend_error');
        return;
      }
      const data = response.data;
      const ajv = new Ajv();
      const validate = ajv.compile(schema);
      const valid = validate(data);
      if (valid) {
        resolve(data);
      } else {
        reject('backend_error');
      }
    })
    .catch((error) => {
      reject('backend_error');
    });
});

const schema = {
  $id: 'http://challenge.monoqi.net/catalog',
  type: 'object',
  definitions: {},
  $schema: 'http://json-schema.org/draft-07/schema#',
  additionalProperties: false,
  properties: {
    articles: {
      $id: '/properties/articles',
      type: 'array',
      items: {
        $id: '/properties/articles/items',
        type: 'object',
        additionalProperties: false,
        properties: {
          sku: {
            $id: '/properties/articles/items/properties/sku',
            type: 'string',
            title: 'The Sku Schema ',
            default: '',
          },
          name: {
            $id: '/properties/articles/items/properties/name',
            type: 'string',
            title: 'The Name Schema ',
            default: '',
          },
          image: {
            $id: '/properties/articles/items/properties/image',
            type: 'string',
            title: 'The Image Schema ',
            default: '',
          },
          price: {
            $id: '/properties/articles/items/properties/price',
            type: 'object',
            additionalProperties: false,
            properties: {
              amount: {
                $id: '/properties/articles/items/properties/price/properties/amount',
                type: 'number',
                title: 'The Amount Schema ',
                default: 0,
              },
              currency: {
                $id: '/properties/articles/items/properties/price/properties/currency',
                type: 'string',
                title: 'The Currency Schema ',
                default: '',
              },
            },
            required: [
              'amount',
              'currency',
            ],
          },
        },
        required: [
          'sku',
          'name',
          'image',
          'price',
        ],
      },
    },
  },
  required: [
    'articles',
  ],
};
