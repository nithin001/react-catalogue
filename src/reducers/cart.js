import Immutable from 'immutable';
const cart = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CART': {
      const items = action.payload.lines;
      return items.reduce((result, item) => {
        return result.set(item.sku, item.quantity);
      }, Immutable.Map());
    }
    case 'ADD_TO_CART': {
      const sku = action.payload;
      if (state.has(sku)) {
        const quantity = state.get(sku);
        return state.set(sku, quantity + 1);
      }
      return state.set(sku, 1);
    }
    case 'REMOVE_FROM_CART': {
      const sku = action.payload;
      const quantity = state.get(sku);
      if (quantity === 1) {
        return state.delete(sku);
      }
      return state.set(sku, quantity - 1);
    }
  }
  return state;
};
export default cart;