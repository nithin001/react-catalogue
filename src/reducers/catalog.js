import Immutable from 'immutable';
const catalog = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CATALOG': {
      const articles = action.payload.articles;
      return Immutable.fromJS(articles);
    }
  }
  return state;
};
export default catalog;