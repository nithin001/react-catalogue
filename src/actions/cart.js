import { get } from '../api/cart';
export const load = () => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_CART_START' });
    return get()
      .then((cart) => {
        dispatch({ type: 'POPULATE_CART', payload: cart });
        dispatch({ type: 'LOAD_CART_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};