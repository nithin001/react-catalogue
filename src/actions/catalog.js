import { get } from '../api/catalog';

export const load = () => (dispatch) => {
  dispatch({ type: 'LOAD_CATALOG_START' });
  return get()
    .then((catalog) => {
      dispatch({ type: 'POPULATE_CATALOG', payload: catalog });
      dispatch({ type: 'LOAD_CATALOG_COMPLETE' });
    })
    .catch(() => {
      dispatch({ type: 'LOAD_ERROR' });
    });
};
