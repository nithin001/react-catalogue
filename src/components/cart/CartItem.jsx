import React from 'react';
import { connect } from 'react-redux';
import { QuantityBar } from './QuantityBar';
import { addToCart, removeFromCart } from '../../actions/cart';
export const CartItem = ({ sku, price, name, quantity, addToCart, removeFromCart }) => {
  return <div className={'cart-item'}>
    <span className={'cart-item__name'}>
      <span className={'cart-item__name cart-item__name--legend'}>Name:&nbsp;</span>
      <span className={'cart-item__name cart-item__name--value'}>{name}</span>
    </span>
    <br/>
    <span className={'cart-item__price cart-item__price--legend'}>Price: &nbsp;
      <span className={'cart-item__price cart-item__price--amount'}>{price.amount}</span>
      &nbsp;
      <span className={'cart-item__price cart-item__price--currency'}>{price.currency}</span>
    </span>
    <QuantityBar sku={sku} quantity={quantity} addToCart={addToCart} removeFromCart={removeFromCart}/>
  </div>;
};

export const mapStateToProps = (state, ownProps) => {
  const article = state.catalog.filter(item => {
    return item.get('sku') === ownProps.sku;
  }).first();
  const price = {
    amount: article.getIn(['price', 'amount']),
    currency: article.getIn(['price', 'currency']),
  };
  return { price, name: article.get('name') };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
    removeFromCart: (sku) => dispatch(removeFromCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(CartItem);
