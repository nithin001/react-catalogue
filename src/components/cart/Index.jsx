import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';
export const Cart = ({ cart }) => {
  const skus = [];
  cart.keySeq().forEach(key => skus.push(key));
  const items = skus.map((key) => (<CartItem
    key={key}
    sku={key}
    quantity={cart.get(key)}
  />));
  return <div className={'cart'}>{items}</div>;
};

export const mapStateToProps = (state) => {
  return { cart: state.cart };
};

export default connect(mapStateToProps, null)(Cart);
