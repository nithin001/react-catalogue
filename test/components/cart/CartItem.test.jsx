import React from 'react';
import { configure, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { CartItem, mapStateToProps } from '../../../src/components/cart/CartItem';
import catalog from '../../../src/reducers/catalog';
configure({ adapter: new Adapter() });

describe('CartItem component', () => {
  it('should render the CartItem with a QuantityBar component', () => {
    const addToCart = jest.fn();
    const removeFromCart = jest.fn();
    const props = {
      sku: '199203',
      quantity: 3,
      price: { amount: 229, currency: 'EUR' },
      name: 'Ipad',
      addToCart,
      removeFromCart
    };
    const cartItem = shallow(<CartItem
      {...props}
    />);
    expect(shallowToJson(cartItem)).toMatchSnapshot();
  });
});

describe('Connected CartItem component', () => {
  it('should return the name of the item in cart', () => {
    var mockCatalog = {
      'articles': [{
        'sku': '199203',
        'name': 'IPad',
        'image': 'https://picsum.photos/480/480/?image=6',
        'price': { 'amount': 229, 'currency': 'EUR' }
      }]
    };
    const catalogState = catalog(undefined, { type: 'POPULATE_CATALOG', payload: mockCatalog });
    const result = mapStateToProps({ catalog: catalogState }, { sku: '199203' });
    expect(result.price).toEqual({ amount: 229, currency: 'EUR' });
    expect(result.name).toEqual('IPad');
  });
});