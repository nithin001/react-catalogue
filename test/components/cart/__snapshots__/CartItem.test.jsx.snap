// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component should render the CartItem with a QuantityBar component 1`] = `
<div
  className="cart-item"
>
  <span
    className="cart-item__name"
  >
    <span
      className="cart-item__name cart-item__name--legend"
    >
      Name: 
    </span>
    <span
      className="cart-item__name cart-item__name--value"
    >
      Ipad
    </span>
  </span>
  <br />
  <span
    className="cart-item__price cart-item__price--legend"
  >
    Price:  
    <span
      className="cart-item__price cart-item__price--amount"
    >
      229
    </span>
     
    <span
      className="cart-item__price cart-item__price--currency"
    >
      EUR
    </span>
  </span>
  <QuantityBar
    addToCart={[MockFunction]}
    quantity={3}
    removeFromCart={[MockFunction]}
    sku="199203"
  />
</div>
`;
